Extract all data acquisition from ResourceService

drewnoakes opened the pull request:
Fixes #1114

This creates a new class `DcpDataSource` that encapsulates all code related to obtaining resource data from DCP in the `ResourceService`. This cleans the service code up a lot and makes the responsibilities clearer.

`Channel<T>` for `KubernetesObject` and `ResourceChange` objects have been replaced with direct code invocations, skipping queuing. We still have a downstream channel for the subscription that ensures front/back are decoupled.

Added some API docs.

## Before

![image](https://github.com/dotnet/aspire/assets/350947/c3318845-676c-49b1-9014-6eab2096d776)

## After

![image](https://github.com/dotnet/aspire/assets/350947/05075fb0-f345-4e39-97be-084671bea9b3)

drewnoakes committed:
Extract all data acquisition from ResourceService

This creates a new class `KubernetesDataSource` that encapsulates all code related to obtaining data from kubernetes in the `ResourceService`. This cleans the service code up a lot and makes the responsibilities clearer.

The use of `Channel<ResourceChange>` has been removed. Instead, code invocations are made directly, skipping queuing. We still have two other channels that ensure front/back are decoupled. The kubernetes channel will ultimately also be removed, leaving just a channel for the subscriber.

Added some API docs.

drewnoakes committed:
Check cancellation token

drewnoakes committed:
Remove Channel<T> for kubernetes objects

This is the second removal of a `Channel<T>` from this component. Now, updates are passed by direct code invocations, skipping queuing. We still have a downstream channel for the subscriber to decouple front/back.

drewnoakes committed:
Inline method

JamesNK commented on src/Aspire.Hosting/Dashboard/ResourceCollection.cs:
```diff
@@ -8,26 +8,16 @@
+/// snapshot and future updates.
 /// </summary>
-internal sealed class ResourceCollection
+internal sealed class ResourceCollection(CancellationToken cancellationToken)
```
I'd like a new name for this. There are two types in hosting called ResourceCollection, and this one isn't a collection.

JamesNK commented on src/Aspire.Hosting/Dashboard/ResourceCollection.cs:
```diff
@@ -37,8 +27,8 @@ public ResourceSubscription Subscribe()
             return new ResourceSubscription(
-                Snapshot: _snapshot.Values.ToList(),
-                Subscription: new ChangeEnumerable(channel, RemoveChannel));
+                Snapshot: [.. _snapshot.Values],
```
Maybe it is just me, but I prefer ye olde `ToList()`.

drewnoakes replied:
I'm happy either way. My preference is that whatever's decided is applied in `.editorconfig` so that we don't get suggestions in the IDE that we don't want, as ignoring them takes a small amount of effort.

JamesNK commented on src/Aspire.Hosting/Dashboard/ResourceCollection.cs:
```diff
@@ -47,30 +37,62 @@ void RemoveChannel(Channel<ResourceChange> channel)
+    /// <param name="resource">The resource that was modified.</param>
+    /// <param name="changeType">The change type (Added, Modified, Deleted).</param>
+    /// <returns>A task that completes when the cache has been updated and all subscribers notified.</returns>
+    public async ValueTask Integrate(ResourceViewModel resource, ObjectChangeType changeType)
```
Could `ResourceViewModel` be renamed to `ResourceModel`? It doesn't have anything to do with views at this point.

drewnoakes replied:
That's coming later. I've been doing the renames in stages. Ultimately it'll become an immutable object and I'm thinking of calling it `ResourceSnapshot`. View models will exist only in the dashboard.

JamesNK commented on src/Aspire.Hosting/Dashboard/ResourceCollection.cs:
```diff
@@ -8,26 +8,16 @@
-        Task.Run(ProcessChanges, cancellationToken);
-    }
-
     public ResourceSubscription Subscribe()
```
Currently the subscribers are the various user's in the dashboard UI (resources, console logs, peer resolver). After the move to gRPC, will it change to each gRPC call being a subscriber?

drewnoakes replied:
No there'll be a single client subscription that fans out to interested parties within the dashboard. You can see more detail in the message I shared on Teams. My current view of the end state is:

![image](https://github.com/dotnet/aspire/assets/350947/5f7a5b54-a792-4cd5-9418-d0649f18b81c)

There's some detail on Teams. If you want to discuss this aspect further -- given it's further out -- please add your thoughts on that thread rather than here.

drewnoakes replied:
End state updated slightly on the client side:

![image](https://github.com/dotnet/aspire/assets/350947/a94a746a-9d9f-4712-9e5d-5ac80502e718)

JamesNK commented on src/Aspire.Hosting/Dashboard/KubernetesDataSource.cs:
```diff
@@ -0,0 +1,593 @@
+    private readonly CancellationToken _cancellationToken;
+    private readonly ILogger _logger;
+
+    private readonly Dictionary<string, Container> _containersMap = [];
+    private readonly Dictionary<string, Executable> _executablesMap = [];
+    private readonly Dictionary<string, Service> _servicesMap = [];
```
I'm wondering how pervasive we want to change the model from inheritance based to composition. Two things are pushing us in the composition direction:
* Generic resources
* gRPC doesn't support inheritance

At what layer does the idea of a generic resource begin?

(note: just posing the question in this PR. I don't think change needs to happen right now)

drewnoakes committed:
Rename ResourceCollection to ResourcePublisher

drewnoakes committed:
Simplify inclusion of docker environment

A few changes here:

- Use more consistent names throughout. They're not "extra" or "additional" arguments. In fact they are a complete replacement. In their absence, the "spec" environment is used.
- Move creation of docker inspection task to view model method, making the various "handle * update" methods more uniform, for future refactoring.
- Remove special handling in `ProcessKubernetesChange` for one scenario. Simplifies the signature, and the code a fair bit.

drewnoakes committed:
Specify list capacity

drewnoakes committed:
Merge ObjectChangeType Add and Modified

We only need to know if the value is "upserted" (updated or inserted), or "deleted".

Recent changes to threading/queuing in the code here changed some timing. The arrival of one kind of resouce can trigger the publication of another, and these would always be "modified", however they could arrive before that resource's stream published that instance. The rest of the code handles these happening out-of-order. We just need to treat add and modified the same way, so they've been merged.

drewnoakes committed:
Use semaphore to serialize kubernetes data processing

There are multiple kubernetes resource types, each with its own monitoring stream. Across these resources, updates arrive concurrently. The update flows for each type of resource can interact with state stored for other resources. Therefore we use a semaphore to ensure that only one resource update can flow through the system at a time.

Another option would have been to make the collections concurrent, or use explicit locking. Concurrent collections are heavy. Explicit locking is tricky to get right. This mutual exclusion via top-level semaphore seems like a safe and elegant approach for now.

drewnoakes committed:
Avoid redundant linear scans

When this method is passed the same collection twice, the `FromSpec` value will always be true, because every item in the list is in the list. Instead we pass `null` and consider all items as from the spec.

drewnoakes committed:
Simplify delegation of projects-as-executables

drewnoakes committed:
Avoid switch on every update

Each instance of `WatchKubernetesResource<T>` produces a single type, that would flow through `ProcessKubernetesChange` to look up the relevant handler. Instead, pass the handler in to `WatchKubernetesResource<T>` so it can be invoked directly.

drewnoakes committed:
Formatting

drewnoakes committed:
Merge remote-tracking branch 'upstream/main' into resource-data-flow

drewnoakes committed:
Rename class to avoid name conflict

This object is a snapshot of a service's state, not a service itself. We have another class which is a service. Append "Snapshot" to differentiate. Future work will extend this concept of snapshots more broadly.

drewnoakes committed:
Replace O(N) scanning with O(1) lookup

drewnoakes committed:
Make resource Endpoints and Services immutable

drewnoakes committed:
Make ResourceViewModel.Environment immutable

Note that it's just the collection that becomes immutable here. The elements are still currently mutable. That will change when we split front/back ends, and have a snapshot on the backend with a view model on the front end.

drewnoakes committed:
Make ContainerViewModel Ports and Args immutable

drewnoakes committed:
Make ExecutableViewModel Arguments immutable

drewnoakes committed:
Make remaining scalar ExecutableViewModel properties immutable

drewnoakes committed:
Make all resource snapshot properties required

drewnoakes committed:
Add API docs for resource snapshot types

karolz-ms commented on src/Aspire.Hosting/Dashboard/KubernetesDataSource.cs:
```diff
@@ -0,0 +1,581 @@
+/// <summary>
+/// Pulls data about kubernetes resources from DCP's kubernetes API. Streams updates to consumers.
+/// </summary>
+internal sealed class KubernetesDataSource
```
Since this is a larger refactoring, maybe it is worth to reflect in class names that this code is not related to vanilla Kubernetes, but to DCP, which has its own set of well-known object kinds. So this class could become `DcpDataSource` for example.

If we don't do it, when we get to adding real Kubernetes support, this will become quite confusing I am afraid.

drewnoakes replied:
Done. However this type is internal and won't ever be public, so we can change the name easily in future if needed.

karolz-ms commented on src/Aspire.Hosting/Dashboard/KubernetesDataSource.cs:
```diff
@@ -0,0 +1,581 @@
+
+        _logger = loggerFactory.CreateLogger<ResourceService>();
+
+        var semaphore = new SemaphoreSlim(1);
```
Any particular reason why this code is using `SemaphoreSlim` vs just ordinary lock?

drewnoakes replied:
So as not to block threads. These are potentially long-running and quite concurrent, especially during startup.

karolz-ms replied:
Oh, I see, in other words, to wait for it asynchronously. Makes sense.

karolz-ms commented on src/Aspire.Hosting/Dashboard/KubernetesDataSource.cs:
```diff
@@ -0,0 +1,581 @@
+        foreach (var ownerReference in endpoint.Metadata.OwnerReferences)
+        {
+            // Find better way for this string switch
+            switch (ownerReference.Kind)
```
DCP model has definitions for well-known kinds: https://github.com/dotnet/aspire/blob/main/src/Aspire.Hosting/Dcp/Model/GroupVersion.cs#L24

drewnoakes replied:
Thanks. They cannot be used in `switch` though as they're not compile-time constants. If you're happy for me to change them from get-only properties to consts, I can use them here. Or we fall back to if/else chaining.

karolz-ms commented on src/Aspire.Hosting/Dashboard/KubernetesDataSource.cs:
```diff
@@ -0,0 +1,581 @@
+        {
+            FillEnvironmentVariables(model.Environment, dockerEnvironment, null);
+        }
+        else if (container.Spec.Env is not null)
```
Containers now have `EffectiveEnv` too https://github.com/dotnet/aspire/issues/1114

drewnoakes replied:
What's the logic here? Should we replace use of `Env` with `EffectiveEnv`, or some other approach?

karolz-ms replied:
Yes, just replacing the use of `Env` with `EffectiveEnv` should do the trick, with the caveat that `Env` is part of the spec, so it is controlled by the AppHost and never changes, vs. `EffectiveEnv` is computed at startup and initially empty.

drewnoakes replied:
I don't see `ContainerSpec.EffectiveEnv` in the code. Perhaps it needs to be mapped in the JSON types. Is it on the spec or the status object? For executables it's on the status.

Perhaps this also means we would no longer need `ComputeEnvironmentVariablesFromDocker` and the process that spins off processes to query the env via the CLI.

drewnoakes replied:
I think I worked this out. Thanks.

drewnoakes committed:
Rename KubernetesDataSource to DcpDataSource

drewnoakes committed:
Use collection literals for empty collections

drewnoakes committed:
Source container variables from status

This data is now available from DCP, so we don't have to launch processes to query docker for this data any more, which simplifies things quite nicely.

drewnoakes committed:
Allocate less memory in ProcessUtil

drewnoakes committed:
Extract duplicate code

drewnoakes committed:
Further deduplicate code

drewnoakes committed:
Make ProjectViewModel derive from ExecutableViewModel

drewnoakes committed:
Remove duplicate razor after consolidating types

drewnoakes committed:
Merge project/executable snapshot construction

Now that project snapshots derive from executable snapshots, we can unify a bunch of construction logic.

drewnoakes committed:
Replace O(N) scanning with O(1) lookup

drewnoakes committed:
Reorder methods

drewnoakes committed:
Start using "snapshot" naming in DCP data source

We still have some types named ViewModel, but they'll be renamed later in the split.

drewnoakes committed:
Rename ObjectChangeType to ResourceChangeType

It's not clear what object the name refers to, so make it more specific. It's internal, so we can always rename it again later. This name pairs nicely with `ObjectChange` too.

drewnoakes committed:
Further consolidate data processing

The methods that handle executable and container updates are largely the same. Extract that commonality to a new method.

tlmii commented on src/Aspire.Hosting/Dashboard/DcpDataSource.cs:
```diff
@@ -0,0 +1,430 @@
+            }
+        }
+
+        if (_resourceAssociatedServicesMap.TryGetValue((resourceKind, name), out var resourceServiceMappings))
```
I'm trying to remember, is the resourceKind part of this mapping because of the original idea that resource names were only guaranteed unique within a particular resource type? But more recently it's been stated that they'll be globally unique. So maybe we could simplify the key for the mapping now?

And by "now" I don't mean this PR since you've already got a lot of changes here. Just making sure someone else reads this thought 😊

drewnoakes replied:
I have wondered the same thing. If we go for unique names, we should make a few other changes throughout. It has an impact on #1274 too, as currently we have a composite key (name/kind) in the protocol.

tlmii reviewed (approved):
This change looks good to me. The diagrams definitely helped set the stage for what was changing, so I apprecaite that. A few questions and comments but nothing major from my perspective.

drewnoakes commented on src/Aspire.Hosting/Dashboard/DcpDataSource.cs:
```diff
@@ -0,0 +1,430 @@
+
+        return new ExecutableViewModel
+        {
+            Name = executable.Metadata.Name,
```
Agreed. Constructors wouldn't much help. We could wrap an executable with a project, but that also feels fairly contrived and has a runtime cost (polymorphism, delegation). This bit of localised duplication seemed like a decent outcome. Open to ideas of course.
